src/main/java/org/sonarsource/scanner/maven/bootstrap/JavaVersionResolver.java

```diff
@@ -0,0 +1,106 @@
+import org.apache.maven.plugin.PluginParameterExpressionEvaluator;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.component.configurator.ComponentConfigurationException;
+import org.codehaus.plexus.component.configurator.converters.basic.StringConverter;
+import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluator;
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.codehaus.plexus.configuration.xml.XmlPlexusConfiguration;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+
+public class JavaVersionResolver {
+  private static final String MAVEN_COMPILER_PLUGIN = "maven-compiler-plugin";
+  private static final String COMPILE_GOAL = MavenUtils.GROUP_ID_APACHE_MAVEN + ":" + MAVEN_COMPILER_PLUGIN + ":compile";
+
+  private final MavenSession session;
+  private final Log log;
+  private List<MojoExecution> mojoExecutions;
+
+  public JavaVersionResolver(MavenSession session, LifecycleExecutor lifecycleExecutor, Log log) {
+    this.session = session;
+    this.log = log;
+    try {
+      this.mojoExecutions = lifecycleExecutor.calculateExecutionPlan(session, true, COMPILE_GOAL).getMojoExecutions();
+    } catch (Exception e) {
+      log.warn(String.format("Failed to get mojo executions for goal '%s': %s", COMPILE_GOAL, e.getMessage()));
+      this.mojoExecutions = new LinkedList<>();
+    }
+  }
+
+  /**
+   * Returns the version of Java used by the maven compiler plugin
+   *
+   * @param pom the project pom
+   * @return the java version
+   */
+  @CheckForNull
+  public String getTarget(MavenProject pom) {
+    return MavenUtils.coalesce(getString("target"), MavenUtils.getPluginSetting(pom, MavenUtils.GROUP_ID_APACHE_MAVEN, MAVEN_COMPILER_PLUGIN, "target", null));
+  }
+
+  @CheckForNull
+  public String getSource(MavenProject pom) {
+    return MavenUtils.coalesce(getString("source"), MavenUtils.getPluginSetting(pom, MavenUtils.GROUP_ID_APACHE_MAVEN, MAVEN_COMPILER_PLUGIN, "source", null));
+
+  }
+
+  @CheckForNull
+  private String getString(String parameter) {
+    try {
+      for (MojoExecution exec : mojoExecutions) {
+        Xpp3Dom configuration = exec.getConfiguration();
+        PlexusConfiguration pomConfiguration = new XmlPlexusConfiguration(configuration);
+        PlexusConfiguration config = pomConfiguration.getChild(parameter);
+
+        ExpressionEvaluator expressionEvaluator = new PluginParameterExpressionEvaluator(session, exec);
+        BasicStringConverter converter = new BasicStringConverter();
+
+        String value = converter.fromExpression(config, expressionEvaluator);
+        if (value != null) {
+          return value;
+        }
+      }
+    } catch (Exception e) {
+      log.warn(String.format("Failed to get parameter '%s' for goal '%s': %s", parameter, COMPILE_GOAL, e.getMessage()));
+    }
+
+    return null;
+  }
+
+  class BasicStringConverter extends StringConverter {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2694)